### ISSUE TITLE
`self` is not resolved to actual class when used as a return type of the contained closure

### ISSUE THREAD

**Issue by weirdan** [vimeo/psalm#316]
Reproduce script (same as in #313)
```php
<?php

class A
{
    /**
     * @var self[]
     */
    private $subitems;
    public function __construct(array $in = [])
    {
        $this->subitems = array_map(function(self $i): self { return $i; }, $in);
    }
}
```
Checking this yields the following false-positive:
```

ERROR: InvalidPropertyAssignment - qqqq.php:11:27 - $this->subitems with declared type 'array<mixed, A>' cannot be assigned type 'array<mixed, self>'
        $this->subitems = array_map(function(self $i): self { return $i; }, $in);
```

**Comment by muglug** [comment]
Thanks!
